pyogrio/_io.pyx

```diff
@@ -2373,15 +2374,18 @@ def ogr_write(
             # TODO: geometry must not be null or errors
             wkb = None if geometry is None else geometry[i]
             if wkb is not None:
-                wkbtype = <int>bytearray(wkb)[1]
-                # may need to consider all 4 bytes: int.from_bytes(wkb[0][1:4], byteorder="little")
-                # use "little" if the first byte == 1
+                wkb_buffer = wkb
+                if wkb_buffer[0] == 1:
+                    # Little endian WKB type.
+                    wkbtype = wkb_buffer[1] + (wkb_buffer[2] << 8) + (wkb_buffer[3] << 16) + (wkb_buffer[4] << 24)
+                else:
+                    # Big endian WKB type.
+                    wkbtype = (wkb_buffer[1] << 24) + (wkb_buffer[2] << 16) + (wkb_buffer[3] << 8) + wkb_buffer[4]
```

Comment:
Due to a subtlety of type promotion in C, a unsigned char shifted 24 position to the left is first cast to a *signed* integer, which can thus cause undefined behavior if the most-significant-bit of this unsigned char is set:

```suggestion
                    wkbtype = wkb_buffer[1] + (wkb_buffer[2] << 8) + (wkb_buffer[3] << 16) + ((unsigned)wkb_buffer[4] << 24)
                else:
                    # Big endian WKB type.
                    wkbtype = ((unsigned)wkb_buffer[1] << 24) + (wkb_buffer[2] << 16) + (wkb_buffer[3] << 8) + wkb_buffer[4]
```

Reply:
Yes, good catch. I didn't think this was possible, but it looks like the types with Z data get the highest bit set.